README.md
[![ci-badge][]][ci] [![license-badge][]][license] [![docs-badge][]][docs] [![rust badge]][rust link]

# aur

`aur` is a package for interacting with the [Arch User Repository] RPC API.

It supports client trait implementations for both asynchronous `hyper` and
synchronous `reqwest`.

### Installation

This library requires at least Rust 1.26.0.

Add the following to your `Cargo.toml`:

```toml
[dependencies]
aur = "~0.1"
```

And the following to your `main.rs` or `lib.rs`:

```rust
extern crate aur;
```

There are two features: `hyper-support` and `reqwest-support`.
`hyper-support` is enabled by default. To enable `reqwest-support`, instead
depend on `aur` like so:

```toml
[dependencies.aur]
default-features = false
features = ["reqwest-support"]
version = "~0.1"
```

### Examples

Asynchronously request information for the `rust-nightly` package:

```rust
extern crate aur;
extern crate hyper;
extern crate hyper_rustls;
extern crate tokio_core;

use aur::bridge::hyper::AurRequester;
use hyper::Client;
use hyper_rustls::HttpsConnector;

let connector = HttpsConnector::new(4);
let client = Client::builder().build(connector);

let done = client.aur_search(Some("rust-nightly"), None).map(|search| {
    assert!(search.result_count >= 2);
}).map_err(|why| {
    println!("Error getting rust-nightly info: {:?}", why);
});

tokio::run(done);
```

Synchronously request information for the `rust-nightly` package:

```rust
extern crate aur;
extern crate reqwest;

use aur::bridge::reqwest::AurRequester;
use reqwest::Client;

let client = Client::new();

let info = client.aur_info(&["rust-nightly"])?;

match info.first() {
    Some(package) => {
        if let Some(ref maintainer) = package.maintainer {
            println!("The package is maintained by: {}", maintainer);
        } else {
            println!("The package has no maintainer");
        }
    },
    None => {
        println!("No package found!");
    },
}
```

### License

[ISC][LICENSE.md].

[ci]: https://travis-ci.org/zeyla/aur.rs
[ci-badge]: https://img.shields.io/travis/zeyla/aur.rs.svg?style=flat-square
[docs]: https://docs.rs/crate/aur
[docs-badge]: https://img.shields.io/badge/docs-online-2020ff.svg?style=flat-square
[LICENSE.md]: https://github.com/zeyla/aur.rs/blob/master/LICENSE.md
[license]: https://opensource.org/licenses/ISC
[license-badge]: https://img.shields.io/badge/license-ISC-blue.svg?style=flat-square
[rust badge]: https://img.shields.io/badge/rust-1.29+-93450a.svg?style=flat-square
[rust link]: https://blog.rust-lang.org/2018/05/10/Rust-1.29.html


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog] and this project adheres to
[Semantic Versioning].

## 0.2.0 - 2018-11-25

### Fixed

- Fix search ([@Morganamilo]) [c:dbe0ff5]

### Changed

- Upgraded reqwest to 0.9 ([@Morganamilo]) [c:7f98d75]
- Increase rustc requirement to 1.29.0 ([@zeyla]) [c:596f4e9]

## 0.1.0 - 2018-04-15

### Added

- Initial release

[@Morganamilo]: https://github.com/Morganamilo

[c:596f4e9]: https://github.com/zeyla/aur.rs/commit/596f4e92cd8dc79c3ca42822a3ee0a8b4521107d
[c:7f98d75]: https://github.com/zeyla/aur.rs/commit/7f98d7511900097aa84502cbb8559f75bbd9c957
[c:dbe0ff5]: https://github.com/zeyla/aur.rs/commit/dbe0ff5879d509fc41be31263cb1407e299ad7fa

[Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: http://semver.org/spec/v2.0.0.html


tests/hyper_support.rs
extern crate aur;
extern crate futures;
extern crate hyper;
extern crate hyper_tls;
extern crate tokio;

use aur::{
    model::SearchBy,
    AurHyperRequester,
};
use futures::Future;
use hyper::client::HttpConnector;
use hyper::{Body, Client};
use hyper_tls::HttpsConnector;

#[inline]
fn client() -> Client<HttpsConnector<HttpConnector>, Body> {
	Client::builder().build(HttpsConnector::new(4).unwrap())
}

#[test]
fn test_info() {
	let done = client().aur_info(&["rust-nightly"]).map(|search| {
        assert_eq!(search.result_count, 1);
    }).map_err(|why| {
        panic!("Err testing info: {:?}", why);
    });

	tokio::run(done);
}

#[test]
fn test_orphans() {
	let done = client().aur_orphans().map(|search| {
        assert!(search.result_count >= 2);
    }).map_err(|why| {
        panic!("Err searching: {:?}", why);
    });

	tokio::run(done);
}

#[test]
fn test_search() {
	let done = client().aur_search("rust").map(|search| {
        assert!(search.result_count >= 2);
    }).map_err(|why| {
        panic!("Err searching: {:?}", why);
    });

	tokio::run(done);
}

#[test]
fn test_search_by() {
	let done = client().aur_search_by("rust", SearchBy::Name).map(|search| {
        assert!(search.result_count >= 2);
    }).map_err(|why| {
        panic!("Err searching: {:?}", why);
    });

	tokio::run(done);
}


src/error.rs
use serde_json::Error as JsonError;
use std::error::Error as StdError;
use std::fmt::{Display, Error as FmtError, Formatter, Result as FmtResult};
use std::result::Result as StdResult;

#[cfg(feature = "http")]
use http::uri::InvalidUri;
#[cfg(feature = "http")]
use http::Error as HttpError;
#[cfg(feature = "hyper")]
use hyper::error::Error as HyperError;
#[cfg(feature = "reqwest")]
use reqwest::{
    Error as ReqwestError,
    Response as ReqwestResponse,
    UrlError as ReqwestUrlError,
};

/// Standard result type for asynchronous functions throughout the library.
pub type Result<T> = StdResult<T, Error>;

/// Enum encompassing the library's possible returned errors.
#[derive(Debug)]
pub enum Error {
    /// An error that occurred while formatting a string.
    Fmt(FmtError),
    /// An error from the `serde_json` crate while deserializing the body of an
    /// HTTP response.
    Json(JsonError),
    /// An error from the `http` crate.
    #[cfg(feature = "http")]
    Http(HttpError),
    /// An error from the `hyper` crate while performing an HTTP request.
    #[cfg(feature = "hyper")]
    Hyper(HyperError),
    /// An error from the `reqwest` crate while performing an HTTP request.
    #[cfg(feature = "reqwest")]
    Reqwest(ReqwestError),
    /// An error indicating a bad request when using `reqwest`.
    #[cfg(feature = "reqwest")]
    ReqwestBad(Box<ReqwestResponse>),
    /// An error indicating an invalid request when using `reqwest`.
    #[cfg(feature = "reqwest")]
    ReqwestInvalid(Box<ReqwestResponse>),
    /// An error indicating a parsing issue when using `reqwest`.
    #[cfg(feature = "reqwest")]
    ReqwestParse(ReqwestUrlError),
    /// An error when building a request's URI from the `http` crate.
    #[cfg(feature = "http")]
    Uri(InvalidUri),
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter) -> FmtResult {
        f.write_str(self.description())
    }
}

impl StdError for Error {
    fn description(&self) -> &str {
        match *self {
            Error::Fmt(ref inner) => inner.description(),
            #[cfg(feature = "http")]
            Error::Http(ref inner) => inner.description(),
            #[cfg(feature = "hyper")]
            Error::Hyper(ref inner) => inner.description(),
            Error::Json(ref inner) => inner.description(),
            #[cfg(feature = "reqwest")]
            Error::Reqwest(ref inner) => inner.description(),
            #[cfg(feature = "reqwest")]
            Error::ReqwestBad(_) => "Request bad",
            #[cfg(feature = "reqwest")]
            Error::ReqwestInvalid(_) => "Request invalid",
            #[cfg(feature = "reqwest")]
            Error::ReqwestParse(ref inner) => inner.description(),
            #[cfg(feature = "http")]
            Error::Uri(ref inner) => inner.description(),
        }
    }
}

impl From<FmtError> for Error {
    fn from(err: FmtError) -> Self {
        Error::Fmt(err)
    }
}

impl From<JsonError> for Error {
    fn from(err: JsonError) -> Self {
        Error::Json(err)
    }
}

#[cfg(feature = "http")]
impl From<HttpError> for Error {
    fn from(err: HttpError) -> Self {
        Error::Http(err)
    }
}

#[cfg(feature = "hyper")]
impl From<HyperError> for Error {
    fn from(err: HyperError) -> Self {
        Error::Hyper(err)
    }
}

#[cfg(feature = "http")]
impl From<InvalidUri> for Error {
    fn from(err: InvalidUri) -> Error {
        Error::Uri(err)
    }
}

#[cfg(feature = "reqwest")]
impl From<ReqwestError> for Error {
    fn from(err: ReqwestError) -> Self {
        Error::Reqwest(err)
    }
}

#[cfg(feature = "reqwest")]
impl From<ReqwestUrlError> for Error {
    fn from(err: ReqwestUrlError) -> Self {
        Error::ReqwestParse(err)
    }
}


src/model.rs
//! Models mapping the API.

use std::fmt;

/// Indicator of how to perform a search.
#[derive(Copy, Clone, Debug)]
pub enum SearchBy {
    /// Indicator to search by name, ascending.
    Name,
    /// Indicator to search by description.
    NameDesc,
    /// Indicator to search by maintainer.
    Maintainer,
    /// Indicator to search by depended packages.
    Depends,
    /// Indicator to search by make-depended packages.
    MakeDepends,
    /// Indicator to search by check-depended packages.
    CheckDepends,
    /// Indicator to search by optionally-depended packages.
    OptDepends,
}

impl fmt::Display for SearchBy {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.write_str(match self {
            SearchBy::Name => "name",
            SearchBy::NameDesc => "name-desc",
            SearchBy::Maintainer => "maintainer",
            SearchBy::Depends => "depends",
            SearchBy::MakeDepends => "makedepends",
            SearchBy::OptDepends => "optdepends",
            SearchBy::CheckDepends => "checkdepends",
        })?;
        Ok(())
    }
}

/// Result data for a search.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Search<T: Send + Sync> {
    /// A list of relevant results.
    pub results: Vec<T>,
    /// The number of results in the [`results`] field.
    ///
    /// [`results`]: #structfield.resultcount
    #[serde(rename = "resultcount")]
    pub result_count: u64,
    /// The type of search that was performed.
    #[serde(rename = "type")]
    pub type_: String,
    /// The version of the API in use.
    pub version: u64,
}

/// A result for a search without additional information metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct SearchResult {
    /// A longer description of the package.
    #[serde(rename = "Description")]
    pub description: Option<String>,
    /// When the package was first submitted.
    #[serde(rename = "FirstSubmitted")]
    pub first_submitted: u64,
    /// The ID of the package.
    #[serde(rename = "ID")]
    pub id: u64,
    /// When the package was last modified.
    #[serde(rename = "LastModified")]
    pub last_modified: u64,
    /// The name of the package's maintainer.
    #[serde(rename = "Maintainer")]
    pub maintainer: Option<String>,
    /// The name of the package.
    #[serde(rename = "Name")]
    pub name: String,
    /// The number of votes that the package has.
    #[serde(rename = "NumVotes")]
    pub num_votes: u64,
    /// When the package was marked as out-of-date.
    #[serde(rename = "OutOfDate")]
    pub out_of_date: Option<u64>,
    /// The name of the base package.
    #[serde(rename = "PackageBase")]
    pub package_base: String,
    /// The ID of the base package.
    #[serde(rename = "PackageBaseID")]
    pub package_base_id: u64,
    /// The relative popularity of the package.
    #[serde(rename = "Popularity")]
    pub popularity: f64,
    /// URL to the package's project home.
    #[serde(rename = "URL")]
    pub url: Option<String>,
    /// Path to the package's snapshot tar.
    #[serde(rename = "URLPath")]
    pub url_path: String,
    /// The version of the package.
    #[serde(rename = "Version")]
    pub version: String,
}

/// A result for a search _with_ additional information metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct InfoResult {
    /// A list of package names that conflicts with this package.
    #[serde(default, rename = "Conflicts")]
    pub conflicts: Vec<String>,
    /// The packages that this package depends upon.
    #[serde(rename = "Depends")]
    pub dependencies: Vec<String>,
    /// A longer description of the package.
    #[serde(rename = "Description")]
    pub description: Option<String>,
    /// When the package was first submitted.
    #[serde(rename = "FirstSubmitted")]
    pub first_submitted: u64,
    /// The ID of the package.
    #[serde(rename = "ID")]
    pub id: u64,
    /// The keywords that the package has been marked with for queryability.
    #[serde(rename = "Keywords")]
    pub keywords: Vec<String>,
    /// When the package was last modified.
    #[serde(rename = "LastModified")]
    pub last_modified: u64,
    /// The license(s) that the project is licensed under.
    #[serde(rename = "License")]
    pub license: Vec<String>,
    /// The name of the package's maintainer.
    #[serde(rename = "Maintainer")]
    pub maintainer: Option<String>,
    /// The list of dependencies to make the package.
    #[serde(default, rename = "MakeDepends")]
    pub make_depends: Vec<String>,
    /// The name of the package.
    #[serde(rename = "Name")]
    pub name: String,
    /// The number of votes that the package has.
    #[serde(rename = "NumVotes")]
    pub num_votes: u64,
    /// The packages that this package optionally depends upon.
    #[serde(default, rename = "OptDepends")]
    pub optional_dependencies: Vec<String>,
    /// When the package was marked as out-of-date.
    #[serde(rename = "OutOfDate")]
    pub out_of_date: Option<u64>,
    /// The name of the base package.
    #[serde(rename = "PackageBase")]
    pub package_base: String,
    /// The ID of the base package.
    #[serde(rename = "PackageBaseID")]
    pub package_base_id: u64,
    /// The relative popularity of the package.
    #[serde(rename = "Popularity")]
    pub popularity: f64,
    /// A list of packages this provides for.
    #[serde(default, rename = "Provides")]
    pub provides: Vec<String>,
    /// URL to the package's project home.
    #[serde(rename = "URL")]
    pub url: Option<String>,
    /// Path to the package's snapshot tar.
    #[serde(rename = "URLPath")]
    pub url_path: String,
    /// The version of the package.
    #[serde(rename = "Version")]
    pub version: String,
}


src/bridge/hyper.rs
//! Bridge to provide a client implementation for the `hyper` crate.
//!
//! # Examples
//!
//! Refer to the documentation for [`AurRequester`].
//!
//! [`AurRequester`]: trait.AurRequester.html

use constants::API_URI;
use futures::{Future, Stream, future};
use hyper::body::Body;
use hyper::client::connect::Connect;
use hyper::client::Client as HyperClient;
use hyper::{Request, Uri};
use model::{InfoResult, Search, SearchBy, SearchResult};
use serde_json;
use std::fmt::{Display, Write};
use std::str::FromStr;
use Error;

macro_rules! ftry {
    ($code:expr) => {
        match $code {
            Ok(v) => v,
            Err(why) => return Box::new(future::err(From::from(why))),
        }
    }
}

/// Trait which defines the methods necessary to interact with the service.
///
/// # Examples
///
/// To bring in the implemenation for the `hyper` Client, simply use the
/// trait:
///
/// ```rust,no_run
/// use aur::AurHyperRequester;
/// ```
///
/// At this point, the methods will be on your Hyper Client.
pub trait AurRequester {
    /// Retrieves information about one or more packages along with metadata.
    ///
    /// # Examples
    ///
    /// Ensure that the `"rust-nightly"` package exists:
    ///
    /// ```rust,ignore
    /// extern crate aur;
    /// extern crate hyper;
    /// extern crate hyper_tls;
    /// extern crate tokio_core;
    ///
    /// use aur::bridge::hyper::AurRequester;
    /// use hyper::Client;
    /// use hyper_tls::HttpsConnector;
    /// use tokio_core::Core;
    ///
    /// let core = Core::new()?;
    ///
    /// let handle = core.handle();
    /// let connector = HttpsConnector::new(4, &handle)?;
    /// let client = Client::configure().connector(connector).build(&handle);
    ///
    /// let done = client.aur_info(&["rust-nightly"]).map(|search| {
    ///     assert_eq!(search.result_count, 1);
    /// }).map_err(|_| ());
    /// ```
    ///
    /// # Errors
    ///
    /// Resolves to [`Error::Fmt`] if there was an error formatting the URI.
    ///
    /// Resolves to [`Error::Hyper`] if there was an error sending the request.
    ///
    /// Resolves to [`Error::Json`] if there was an error deserializing the
    /// response body.
    ///
    /// Resolves to [`Error::Uri`] if there was an error parsing the Uri.
    ///
    /// [`Error::Fmt`]: ../../enum.Error.html#variant.Fmt
    /// [`Error::Hyper`]: ../../enum.Error.html#variant.Hyper
    /// [`Error::Json`]: ../../enum.Error.html#variant.Json
    /// [`Error::Uri`]: ../../enum.Error.html#variant.Uri
    fn aur_info<T: Display>(&self, packages: &[T])
        -> Box<Future<Item = Search<InfoResult>, Error = Error> + Send>;

    /// Searches for packages by a query, searching by maintainer name.
    ///
    /// # Examples
    ///
    /// Ensure that at least two packages return for the `"rust"` query,
    /// searching by name:
    ///
    /// ```rust,ignore
    /// extern crate aur;
    /// extern crate hyper;
    /// extern crate hyper_tls;
    /// extern crate tokio_core;
    ///
    /// use aur::{
    ///     bridge::hyper::AurRequester,
    ///     model::SearchBy,
    /// };
    /// use hyper::Client;
    /// use hyper_tls::HttpsConnector;
    /// use tokio_core::Core;
    ///
    /// let handle = core.handle();
    /// let connector = HttpsConnector::new(4, handle)?;
    /// let client = Client::configure().connector(connector).build(&handle);
    ///
    /// let done = client.aur_search("rust", SearchBy::Name).map(|search| {
    ///     assert!(search.result_count >= 2);
    /// }).map_err(|_| ());
    /// ```
    ///
    /// # Errors
    ///
    /// Resolves to [`Error::Fmt`] if there was an error formatting the URI.
    ///
    /// Resolves to [`Error::Hyper`] if there was an error sending the request.
    ///
    /// Resolves to [`Error::Json`] if there was an error deserializing the
    /// response body.
    ///
    /// Resolves to [`Error::Uri`] if there was an error parsing the Uri.
    ///
    /// [`Error::Fmt`]: ../../enum.Error.html#variant.Fmt
    /// [`Error::Hyper`]: ../../enum.Error.html#variant.Hyper
    /// [`Error::Json`]: ../../enum.Error.html#variant.Json
    /// [`Error::Uri`]: ../../enum.Error.html#variant.Uri
    fn aur_search_by(&self, query: &str, by: SearchBy)
        -> Box<Future<Item = Search<SearchResult>, Error = Error> + Send>;

    /// Search for packages by a query.
    ///
    /// # Examples
    ///
    /// Ensure that at least two packages return for the `"rust"` query, not
    /// specifying a maintainer:
    ///
    /// ```rust,ignore
    /// extern crate aur;
    /// extern crate hyper;
    /// extern crate hyper_tls;
    /// extern crate tokio_core;
    ///
    /// use aur::bridge::hyper::AurRequester;
    /// use hyper::Client;
    /// use hyper_tls::HttpsConnector;
    /// use tokio_core::Core;
    ///
    /// let handle = core.handle();
    /// let connector = HttpsConnector::new(4, handle)?;
    /// let client = Client::configure().connector(connector).build(&handle);
    ///
    /// let done = client.aur_search(Some("rust"), None).map(|search| {
    ///     assert!(search.result_count >= 2);
    /// }).map_err(|_| ());
    /// ```
    ///
    /// # Errors
    ///
    /// Resolves to [`Error::Fmt`] if there was an error formatting the URI.
    ///
    /// Resolves to [`Error::Hyper`] if there was an error sending the request.
    ///
    /// Resolves to [`Error::Json`] if there was an error deserializing the
    /// response body.
    ///
    /// Resolves to [`Error::Uri`] if there was an error parsing the Uri.
    ///
    /// [`Error::Fmt`]: ../../enum.Error.html#variant.Fmt
    /// [`Error::Hyper`]: ../../enum.Error.html#variant.Hyper
    /// [`Error::Json`]: ../../enum.Error.html#variant.Json
    /// [`Error::Uri`]: ../../enum.Error.html#variant.Uri
    fn aur_search(&self, query: &str)
        -> Box<Future<Item = Search<SearchResult>, Error = Error> + Send>
    {
        self.aur_search_by(query, SearchBy::NameDesc)
    }

    /// Search for a list of orphaned packages.
    ///
    /// # Examples
    ///
    /// Retrieve a list of orphaned packages:
    ///
    /// ```rust,ignore
    /// extern crate aur;
    /// extern crate hyper;
    /// extern crate hyper_tls;
    /// extern crate tokio_core;
    ///
    /// use aur::bridge::hyper::AurRequester;
    /// use hyper::Client;
    /// use hyper_tls::HttpsConnector;
    /// use tokio_core::Core;
    ///
    /// let handle = core.handle();
    /// let connector = HttpsConnector::new(4, handle)?;
    /// let client = Client::configure().connector(connector).build(&handle);
    ///
    /// let done = client.aur_orphans().map(|search| {
    ///     println!("Orphaned packages: {}", search.result_count);
    /// }).map_err(|_| ());
    /// ```
    ///
    /// # Errors
    ///
    /// Resolves to [`Error::Fmt`] if there was an error formatting the URI.
    ///
    /// Resolves to [`Error::Hyper`] if there was an error sending the request.
    ///
    /// Resolves to [`Error::Json`] if there was an error deserializing the
    /// response body.
    ///
    /// Resolves to [`Error::Uri`] if there was an error parsing the Uri.
    ///
    /// [`Error::Fmt`]: ../../enum.Error.html#variant.Fmt
    /// [`Error::Hyper`]: ../../enum.Error.html#variant.Hyper
    /// [`Error::Json`]: ../../enum.Error.html#variant.Json
    /// [`Error::Uri`]: ../../enum.Error.html#variant.Uri
    fn aur_orphans(&self)
        -> Box<Future<Item = Search<SearchResult>, Error = Error> + Send>
    {
        self.aur_search_by("", SearchBy::Maintainer)
    }
}

impl<C> AurRequester for HyperClient<C, Body>
    where C: Connect + Sync + 'static,
          C::Future: 'static,
          C::Transport: 'static {
    fn aur_info<T: Display>(&self, packages: &[T])
        -> Box<Future<Item = Search<InfoResult>, Error = Error> + Send + 'static> {
        let mut url = format!("{}&type=info", API_URI);

        for package in packages {
            if let Err(why) = write!(url, "&arg[]={}", package) {
                return Box::new(future::err(Error::Fmt(why)));
            }
        }

        let c = &url;
        let uri = ftry!(Uri::from_str(c));

        let mut request = Request::get(uri);
        let req = ftry!(request.body(Body::empty()));

        Box::new(self.request(req)
            .and_then(|res| res.into_body().concat2())
            .map_err(From::from)
            .and_then(|body| serde_json::from_slice(&body).map_err(From::from)))
    }

    fn aur_search_by(&self, query: &str, by: SearchBy)
        -> Box<Future<Item = Search<SearchResult>, Error = Error> + Send + 'static> {
        let url = format!("{}&type=search&arg={}&by={}", API_URI, query, by);
        let c = &url;
        let uri = ftry!(Uri::from_str(c));

        let mut request = Request::get(uri);
        let req = ftry!(request.body(Body::empty()));

        Box::new(self.request(req)
            .and_then(|res| res.into_body().concat2())
            .map_err(From::from)
            .and_then(|body| serde_json::from_slice(&body).map_err(From::from)))
    }
}


src/lib.rs
//! [![ci-badge][]][ci] [![license-badge][]][license] [![docs-badge][]][docs] [![rust badge]][rust link]
//!
//! # aur
//!
//! `aur` is a package for interacting with the [Arch User Repository] RPC API.
//!
//! It supports client trait implementations for both asynchronous `hyper` and
//! synchronous `reqwest`.
//!
//! ### Installation
//!
//! Add the following to your `Cargo.toml`:
//!
//! ```toml
//! [dependencies]
//! aur = "~0.1"
//! ```
//!
//! And the following to your `main.rs` or `lib.rs`:
//!
//! ```rust
//! extern crate aur;
//! ```
//!
//! There are two features: `hyper-support` and `reqwest-support`.
//! `hyper-support` is enabled by default. To enable `reqwest-support`, instead
//! depend on `aur` like so:
//!
//! ```toml
//! [dependencies.aur]
//! default-features = false
//! features = ["reqwest-support"]
//! version = "~0.1"
//! ```
//!
//! ### Examples
//!
//! Asynchronously request information for the `rust-nightly` package:
//!
//! ```rust
//! # #[cfg(feature = "tokio")]
//! mod inner {
//! extern crate aur;
//! extern crate hyper;
//! extern crate hyper_tls;
//! extern crate tokio;
//!
//! #     fn main() -> Result<(), Box<::std::error::Error>> {
//! #
//! use aur::bridge::hyper::AurRequester;
//! use hyper::Client;
//! use hyper::net::HttpsConnector;
//! use hyper_tls::NativeTlsClient;
//!
//! let connector = HttpsConnector::new(4);
//! let client = Client::builder().build(connector);
//!
//! let done = client.aur_search(Some("rust-nightly"), None).map(|search| {
//!     assert!(search.result_count >= 2);
//! }).map_err(|why| {
//!     println!("Error getting rust-nightly info: {:?}", why);
//! });
//!
//! tokio::run(done);
//! #
//! #         Ok(())
//! #     }
//! # }
//! #
//! # #[cfg(feature = "tokio")]
//! # fn main() {
//! #     inner::main().unwrap();
//! # }
//! #
//! # #[cfg(not(feature = "tokio"))]
//! # fn main() {}
//! ```
//!
//! Synchronously request information for the `rust-nightly` package:
//!
//! ```rust
//! # #[cfg(feature = "reqwest")]
//! extern crate aur;
//! # #[cfg(feature = "reqwest")]
//! extern crate reqwest;
//!
//! # #[cfg(feature = "reqwest")]
//! # fn try_main() -> Result<(), Box<::std::error::Error>> {
//! #
//! use aur::bridge::reqwest::AurRequester;
//! use reqwest::Client;
//!
//! let client = Client::new();
//!
//! let info = client.aur_info(&["rust-nightly"])?;
//!
//! match info.results.first() {
//!     Some(package) => {
//!         if let Some(ref maintainer) = package.maintainer {
//!             println!("The package is maintained by: {}", maintainer);
//!         } else {
//!             println!("The package has no maintainer");
//!         }
//!     },
//!     None => {
//!         println!("No package found!");
//!     },
//! }
//! #     Ok(())
//! # }
//! #
//! # #[cfg(not(feature = "reqwest"))]
//! # fn try_main() -> Result<(), Box<::std::error::Error>> { Ok(()) }
//! #
//! # fn main() {
//! #     try_main().unwrap();
//! # }
//! ```
//!
//! ### License
//!
//! ISC.
//!
//! [Arch User Repository]: https://aur.archlinux.org/
//! [ci]: https://travis-ci.org/zeyla/aur.rs
//! [ci-badge]: https://img.shields.io/travis/zeyla/aur.rs.svg?style=flat-square
//! [docs]: https://docs.rs/crate/aur
//! [docs-badge]: https://img.shields.io/badge/docs-online-2020ff.svg?style=flat-square
//! [LICENSE.md]: https://github.com/zeyla/aur.rs/blob/master/LICENSE.md
//! [license]: https://opensource.org/licenses/ISC
//! [license-badge]: https://img.shields.io/badge/license-ISC-blue.svg?style=flat-square
//! [rust badge]: https://img.shields.io/badge/rust-1.29+-93450a.svg?style=flat-square
//! [rust link]: https://blog.rust-lang.org/2018/05/10/Rust-1.29.html
#![deny(missing_docs)]

#[macro_use] extern crate serde_derive;

extern crate serde;
extern crate serde_json;

#[cfg(feature = "futures")]
extern crate futures;
#[cfg(feature = "http")]
extern crate http;
#[cfg(feature = "hyper")]
extern crate hyper;
#[cfg(feature = "reqwest")]
extern crate reqwest;

pub mod bridge;
pub mod model;

mod constants;
mod error;

pub use error::{Error, Result};

#[cfg(feature = "hyper")]
pub use self::bridge::hyper::AurRequester as AurHyperRequester;
#[cfg(feature = "reqwest")]
pub use self::bridge::reqwest::AurRequester as AurReqwestRequester;
